barcode_library_characterizer/library_quality_control.py

```diff
@@ -135,13 +135,50 @@ def build_hamming_distance_multithreading(
             row[1]
             for row in reader ]
 
-    pairs = [ (slow,fast) for slow in range(len(barcodes))
-              for fast in range(slow+1, len(barcodes)) ]
-
-    for slow,fast in pairs:
-        distance = hamming_distance(barcodes[fast], barcodes[slow])
-        hamming_histogram[distance] += 1
+    return build_histogram_simple(barcodes)
+        
+        
+def build_histogram_simple(barcodes):
+    arr = [0]*(len(barcodes[0])+1)
+    c = Counter( hamming_distance(barcodes[fast], barcodes[slow])
+                    for slow,fast in pairs(len(barcodes)) )
+    for i,count in c.items():
+        arr[i] = count
+    return arr
+
+def build_histogram_arr(barcodes):
+    arr = [0]*(len(barcodes[0])+1)
+
+    for slow,fast in pairs(len(barcodes)):
+        d = hamming_distance(barcodes[fast], barcodes[slow])
+        arr[d] += 1
+    return arr
+
+# Trick to share state between processes, add it to a module field 
+from . import shared
+
+# can't define as a local function because it can't be pickled to send between processes
+def update(i, j):
+    d = hamming_distance(shared.barcodes[i], shared.barcodes[j])
+    shared.A[d] += 1
+    #return d
+
+# https://stackoverflow.com/a/1721911
+def initProcess(A, barcodes):
+    shared.A = A
+    shared.barcodes = barcodes
+
+
+def build_histogram_multiprocess(barcodes):
+    arr = [0]*(len(barcodes[0])+1)
+    A = Array('i', arr)
+
+    with Pool(initializer=initProcess,initargs=(A,barcodes)) as p:
+        p.starmap(update, pairs(len(barcodes)))

```

Comment:
Nice! Was the previous option not working or slow because every thresd needed to access the same dictionary?

Maybe it's even easier if we have one Counter per thread and the Counters can be summed at the end. This way there's no sharing between threads.

Reply:
What do you mean with "previous option"? The implementation of `build_hamming_distance_histogram_parallel`? 
I wasn't able to run that as it was coupled with the csv read and I didn't want to break it apart. 
Some differences are:
- update of defaultdict done synchronously  on main thread vs update of shared Array on each process
- passing barcodes pairs vs passing indexes and lookup on shared array (I'm not sure how to test what is being serialized and transferred between processes)